fawltydeps/main.py

```diff
@@ -220,56 +220,67 @@ def print_json(self, out: TextIO) -> None:
-            if details:
+        def render_imports() -> Iterator[str]:
+            if detailed:
                 # Sort imports by source, then by name
                 for imp in sorted(self.imports, key=attrgetter("source", "name")):
-                    print(f"{imp.source}: {imp.name}", file=out)
+                    yield f"{imp.source}: {imp.name}"
             else:
                 unique_imports = {i.name for i in self.imports}
-                print("\n".join(sorted(unique_imports)), file=out)
-
-        if self.is_enabled(Action.LIST_DEPS):
-            if details:
-                # Sort dependencies by location, then by name
-                for dep in sorted(
-                    set(self.declared_deps), key=attrgetter("source", "name")
-                ):
-                    print(f"{dep.source}: {dep.name}", file=out)
+                yield from sorted(unique_imports)
+
+        def render_declared_deps() -> Iterator[str]:
+            if detailed:
+                # Sort dependencies by source, then by name
+                unique_deps = set(self.declared_deps)
+                for dep in sorted(unique_deps, key=attrgetter("source", "name")):
+                    yield f"{dep.source}: {dep.name}"
             else:
-                print(
-                    "\n".join(sorted(set(d.name for d in self.declared_deps))), file=out
-                )
+                yield from sorted(set(d.name for d in self.declared_deps))
 
-        if self.is_enabled(Action.REPORT_UNDECLARED) and self.undeclared_deps:
-            print("\nThese imports appear to be undeclared dependencies:", file=out)
+        def render_undeclared() -> Iterator[str]:
+            yield "\nThese imports appear to be undeclared dependencies:"
             for undeclared in self.undeclared_deps:
-                print(f"- {undeclared.render(details)}", file=out)
+                yield f"- {undeclared.render(detailed)}"
 
-        if self.is_enabled(Action.REPORT_UNUSED) and self.unused_deps:
-            print(f"\n{UNUSED_DEPS_OUTPUT_PREFIX}:", file=out)
+        def render_unused() -> Iterator[str]:
+            yield f"\n{UNUSED_DEPS_OUTPUT_PREFIX}:"
             for unused in sorted(self.unused_deps, key=lambda d: d.name):
-                print(f"- {unused.render(details)}", file=out)
+                yield f"- {unused.render(detailed)}"
+
+        def output(lines: Iterator[str]) -> None:
+            for line in lines:
+                print(line, file=out)
+
+        if self.is_enabled(Action.LIST_SOURCES):
+            output(render_sources())
+        if self.is_enabled(Action.LIST_IMPORTS):
+            output(render_imports())
+        if self.is_enabled(Action.LIST_DEPS):
+            output(render_declared_deps())
+        if self.is_enabled(Action.REPORT_UNDECLARED) and self.undeclared_deps:
+            output(render_undeclared())
+        if self.is_enabled(Action.REPORT_UNUSED) and self.unused_deps:
+            output(render_unused())
```

Comment:
This might be more readable but I don't see why Pylint is happier with it. The number of `if`-branches stays the same :thinking:

Reply:
Yes, but they're spread over different (nested) helper functions, which I suspect appeases Pylint in the end. Specifically, the `if detailed: ... else: ...` blocks are now neatly organized into the helper functions, and I have to say I actually find the current organization quite readable (once you get over the 6 nested functions inside this method 😅 )